internal/provider/aws/provider_test.go

```diff
@@ -0,0 +1,4154 @@
+								},
+								Groups: []types.GroupIdentifier{
+									{
+										GroupId:   aws.String(groupID),
+										GroupName: aws.String(sgName),
+									},
+									{
+										GroupId:   aws.String("sg-01"),
+										GroupName: aws.String("sg-01"),
+									},
+									{
+										GroupId:   aws.String("sg-02"),
+										GroupName: aws.String("sg-02"),
+									},
+								},
+							},
+						},
+					}, nil
+				})
+
+			mockec2Client.EXPECT().
+				ModifyNetworkInterfaceAttribute(ctx, gomock.AssignableToTypeOf(&ec2.ModifyNetworkInterfaceAttributeInput{})).
+				DoAndReturn(func(_ context.Context, input *ec2.ModifyNetworkInterfaceAttributeInput, _ ...func(*ec2.Options)) (*ec2.ModifyNetworkInterfaceAttributeOutput, error) {
+					switch aws.ToString(input.NetworkInterfaceId) {
+					case eniID:
+						Expect(input.Groups).To(ConsistOf("sg-01", "sg-02"))
+						return &ec2.ModifyNetworkInterfaceAttributeOutput{}, nil
+					case eni01ID:
+						Expect(input.Groups).To(ConsistOf("sg-01", "sg-02"))
+						return &ec2.ModifyNetworkInterfaceAttributeOutput{}, nil
+					default:
+						return nil, fmt.Errorf("unexpected interface ID: %s", aws.ToString(input.NetworkInterfaceId))
+					}
+				}).
+				Times(2)
+			mockec2Client.EXPECT().
+				DeleteSecurityGroup(ctx, gomock.AssignableToTypeOf(&ec2.DeleteSecurityGroupInput{})).
+				DoAndReturn(func(_ context.Context, input *ec2.DeleteSecurityGroupInput, _ ...func(*ec2.Options)) (*ec2.DeleteSecurityGroupOutput, error) {
+					Expect(aws.ToString(input.GroupId)).To(Equal(groupID))
+					return &ec2.DeleteSecurityGroupOutput{}, fmt.Errorf("delete security group error")
+				})
+
+			err := p.ReconcileFirewallRulesDeletion(ctx, log, nodeName, "")
+			Expect(err).To(MatchError(&provider.Error{
+				Code: "InternalError",
+				Msg:  "failed to delete security group: delete security group error",
+			}))
+		})
+
+		It("should return nil when no aws api call returns an error", func() {
+			mockec2Client.EXPECT().
+				DescribeSecurityGroups(ctx, gomock.AssignableToTypeOf(&ec2.DescribeSecurityGroupsInput{})).
+				DoAndReturn(func(_ context.Context, input *ec2.DescribeSecurityGroupsInput, _ ...func(*ec2.Options)) (*ec2.DescribeSecurityGroupsOutput, error) {
+					Expect(input.Filters).To(ConsistOf(filters))
+					return &ec2.DescribeSecurityGroupsOutput{
+						SecurityGroups: []types.SecurityGroup{
+							{
+								GroupId:   aws.String(groupID),
+								GroupName: aws.String(sgName),
+							},
+						},
+					}, nil
+				})
+			mockec2Client.EXPECT().
+				DescribeNetworkInterfaces(ctx, gomock.AssignableToTypeOf(&ec2.DescribeNetworkInterfacesInput{})).
+				DoAndReturn(func(_ context.Context, input *ec2.DescribeNetworkInterfacesInput, _ ...func(*ec2.Options)) (*ec2.DescribeNetworkInterfacesOutput, error) {
+					Expect(input.Filters).To(HaveLen(1))
+					Expect(aws.ToString(input.Filters[0].Name)).To(Equal("group-id"))
+					Expect(input.Filters[0].Values).To(ConsistOf(groupID))
```

Comment:
Idem, you forgot about https://github.com/quortex/kubestatic/pull/186#discussion_r2060264221